lib/Web/Server.hs

```diff
@@ -42,6 +47,89 @@ url = BaseUrl Http "127.0.0.1" 8200 ""
+throwCreateError = \case
+  CRSuccess{} -> throwError err500 { errBody = "Expected create error, got CRSuccess." }
+  CRCreateError createError ->
+    case createError of
+      res@CEParentDirectoryIsEntry{} -> throwError err409 { errBody = cofferServerErrorToBytes (pretty res) 200 }
+      res@CEDestinationIsDirectory{} -> throwError err409 { errBody = cofferServerErrorToBytes (pretty res) 201 }
+      res@CEEntryAlreadyExists{} -> throwError err409 { errBody = cofferServerErrorToBytes (pretty res) 202 }
+
+throwSetFieldError :: SetFieldResult -> Handler a
+throwSetFieldError = \case
+  SFRSuccess{} -> throwError err500 { errBody = "Expected set-field error, got SFRSuccess." }
+  res@SFREntryNotFound{} -> throwError err404 { errBody = cofferServerErrorToBytes (pretty res) 300 }
+  res@SFRMissingFieldContents{} -> throwError err400 { errBody = cofferServerErrorToBytes (pretty res) 301 }
+
+throwDeleteFieldError :: DeleteFieldResult -> Handler a
+throwDeleteFieldError = \case
+  DFRSuccess{} -> throwError err500 { errBody = "Expected delete-field error, got DFRSuccess." }
+  res@DFREntryNotFound{} -> throwError err404 { errBody = cofferServerErrorToBytes (pretty res) 400 }
+  res@DFRFieldNotFound{} -> throwError err404 { errBody = cofferServerErrorToBytes (pretty res) 401 }
+
+throwCopyOrRenameError :: Bool -> CopyResult -> Handler a
+throwCopyOrRenameError rename = \case
+  CPRSuccess{} -> throwError err500 { errBody = "Expected " <> cmdName <> " error, got CPRSuccess." }
+  res@CPRPathNotFound{} -> throwError err404 { errBody = cofferServerErrorToBytes (prettySingleMessage res) 500 }
+  res@CPRMissingEntryName{} -> throwError err400 { errBody = cofferServerErrorToBytes (prettySingleMessage res) 501 }
+  res@CPRSamePath{} -> throwError err409 { errBody = cofferServerErrorToBytes (prettySingleMessage res) 502 }
+  res@(CPRCreateErrors createErrors) -> do
+    let messages = buildCopyOrRenameResult rename res <&> pretty
+    let codes = createErrors <&> createErrorToCode . snd
+    let serverErrors = zipWith CofferServerError messages codes
+    let body = encodingToLazyByteString $ toEncoding serverErrors
+    throwError err409 { errBody = body }
+  where
+    cmdName = if rename then "rename" else "copy"
+
+    prettySingleMessage :: CopyResult -> Text
+    prettySingleMessage res =
+      buildCopyOrRenameResult rename res
+        & unlinesF
+        & pretty
+
+    createErrorToCode :: CreateError -> Int
+    createErrorToCode = \case
+      CEParentDirectoryIsEntry{} -> 503
+      CEDestinationIsDirectory{} -> 504
+      CEEntryAlreadyExists{} -> 505
+
+throwDeleteError :: DeleteResult -> Handler a
+throwDeleteError = \case
+  DRSuccess{} -> throwError err500 { errBody = "Expected delete error, got DRSuccess." }
+  res@DRPathNotFound{} -> throwError err404 { errBody = cofferServerErrorToBytes (pretty res) 600 }
+  res@DRDirectoryFound{} -> throwError err404 { errBody = cofferServerErrorToBytes (pretty res) 601 }
```

Comment:
This one's not a 404. But I'm not sure what the appropriate status code should be here either...
Hmm let's just default to 400.

Reply:
Done